edb/schema/pointers.py

```diff
@@ -1545,6 +1560,80 @@ def _cmd_tree_from_ast(
         return cmd
 
 
+class AlterPointerUpperCardinality(
+    referencing.ReferencedInheritingObjectCommand[Pointer_T],
+    inheriting.AlterInheritingObjectFragment[Pointer_T],
+    sd.AlterSpecialObjectField[Pointer_T],
+    PointerCommandOrFragment[Pointer_T],
+):
+    """Handler for the "cardinality" field changes."""
+
+    def _alter_begin(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> s_schema.Schema:
+        orig_schema = schema
+        schema = super()._alter_begin(schema, context)
+        scls = self.scls
+
+        orig_card = scls.get_cardinality(orig_schema)
+        new_card = scls.get_cardinality(schema)
+
+        if orig_card == new_card:
+            # The actual value hasn't changed, nothing to do here.
+            return schema
+
+        # Cardinality alters of pointers used in expressions is prohibited.
+        # Eventually we may be able to relax this by revalidating the
+        # expressions and making sure their meaning and cardinality does not
+        # change.
+        vn = scls.get_verbosename(schema, with_parent=True)
+        schema = self._prohibit_if_expr_refs(
```

Comment:
Hm, why does it need to be prohibit instead of propagate? (The comment in in the propagate call for SetPointerType is wrong, and we don't actually prohibit type changes; I think that we do actually propagate them correctly now and generate errors and updates where necessary?)

Reply:
I'm not 100% sure what the correct approach to propagation should be.  Should cardinality changes automatically affect cardinality of computable expressions that might use this pointer?  IMO, we should only do autopropagation if referrers are unaffected, and require an explicit opt-in otherwise.

Reply:
Actually I think I might be wrong here, since we won't actually do all the updates necessary in actual schema stuff for things like functions; some of that can be fixed just by setting `metadata_only=False`, though

Reply:
My inclination would be to automatically propagate the changes, though an opt-in would probably work fine too? Do you have syntax ideas for that?

Reply:
> And then what should ALTER TYPE Foo ALTER PROPERTY bar SET DEFAULT := 'new default' do? Should it propagate to subtypes even though they have an explicit override set? Probably not.

I'm not sure I follow this example. The prevalent mode of operation for edgedb users will be modifying their schema via SDL. In which case the user would just change your schema to 

```
type Foo {
    property bar -> str { default := 'new default' }
}

type Sub extending Foo {
    overloaded property bar -> str { default := 'spam' }
}
```

And we can compile this to a targeted DDL that would just update `Foo.bar`. No need for auto-propagation or for CASCADE/RESTRICT modifiers.  Basically our DDL should be robust enough to support SDL and migrations, and I don't think we need to pack more features in it than what's absolutely required.

Reply:
> I'm not sure I follow this example. The prevalent mode of operation for edgedb users will be modifying their schema via SDL.

The `CASCADE` thing is hypothetical, I didn't mean that it is necessary for this PR.